changes/2023_7_12_update-keystore-structure/background.md

```diff
@@ -0,0 +1,286 @@
+while possible to disambiguate, this a very complicated problem.
+It would be much better if the system was correct by construction and there could only be a single `active` version.
+
+# Requirements
+
+1. Branch key ids and versions creation is deterministically unique
+
+1. Additional hooks to authorize access to branch keys
+
+1. An optional branch key id added to the create input
+
+# Out of Scope
+
+- Migration from the developer preview version
+- Local policy evaluation for use of branch keys
+
+# Design questions
+
+_Preferred options are identified like this._
+
+## How can duplicate branch keys be rejected by construction?
+
+### _Make the `active` version identifier part of the sort key._
+
+This would mean that there is a single active sort key for a branch key.
+Because the partition key/sort key is now a unique value this means that dynamoDB enforces uniqueness naturally.
+By enforcing a consistent write that has a conditional update that requires that this partition/sort key not exist this enforces both uniqueness and creation,
+as opposed to update of an existing branch key.
+
+### Query the table for the existing branch key
+
+Here we would query only the partition key.
+By only returning the partition key and sort keys we can can be comfortable that we will always return all records in every successful case.
+This is because 1 DDB read unit 4K, much more than ~200 bytes for 6 GUIDs (3 branch key/version records).
+
+However, this function has a fundamental race.
+Even if the read is strongly consistent, while this means that there will be a winner, this still means that we have created duplicate records.
+The system will eventually recover, but there exists a possibility that some other caller may encrypt values.
+Since the hierarchy has a cache, this problem may not even be small.
+
+It is significantly better for the request to be as atomic as possible.
+
+## How do we distinguish active from non-active versions?
+
+### _Insert 2 records, the active record and the version record_
+
+Here we insert the version record and that data is immutable.
+Then we also insert a “version” that is the active version.
+Having an immutable record is a nice property.
+The downside here is that there are 2 records to insert.
+The other advantage here is that getting this record by version is easy.
+This is important because on decrypt we get the branch key by version and not by “active”.
+
+### ReEncrypt the active version to no longer be active
+
+Here we would insert a single record and then reEncrypt this value to be no longer active.
+This means that the transaction to update is more complicated.
+I think that it is still possible to update everything, but now there is a single copy of the branch key that is getting updated.
+It seems impossible, but not ever updating a row in the database is the safest way to go.
+Especially if there is ever a need to recover, this record always exists.
+
+## How do we create these two ciphertext blobs?
```

Comment:
From context I'm inferring that `ciphertext blobs` refers to the `enc` attribute on the keystore records, but it would be great if you could make it clearer that that's what you're referring to... and maybe explain what the problem is here (i.e. why do we need to give special consideration to how we generate these blobs)